controller/uninstall_controller.go

```diff
@@ -0,0 +1,468 @@
+		stopCh:                    stopCh,
+		volumeInformerSynced:      volumeInformer.Informer().HasSynced,
+		engineInformerSynced:      engineInformer.Informer().HasSynced,
+		replicaInformerSynced:     replicaInformer.Informer().HasSynced,
+		engineImageInformerSynced: engineImageInformer.Informer().HasSynced,
+		nodeInformerSynced:        nodeInformer.Informer().HasSynced,
+		daemonSetInformerSynced:   daemonSetInformer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		), "longhorn-uninstall"),
+	}
+
+	volumeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineInformer.Informer().AddEventHandler(c.controlleeHandler())
+	replicaInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineImageInformer.Informer().AddEventHandler(c.controlleeHandler())
+	nodeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	daemonSetInformer.Informer().AddEventHandler(c.controlleeHandler())
+	return c
+}
+
+func (c *UninstallController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *UninstallController) enqueueControlleeChange() {
+	c.queue.AddRateLimited("uninstall")
+}
+
+func (c *UninstallController) Run() error {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	if !controller.WaitForCacheSync("longhorn uninstall", c.stopCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+
+		return errors.New("Failed to sync informers")
+	}
+
+	if err := c.checkPreconditions(); err != nil {
+		c.stop()
+		return err
+	}
+
+	startTime := time.Now()
+	logrus.Infof("Uninstalling...")
+	defer func() {
+		if c.stopErr == nil {
+			fields := logrus.Fields{"runtime": time.Now().Sub(startTime)}
+			logrus.WithFields(fields).Infof("Uninstallation completed.")
+		}
+	}()
+	go wait.Until(c.worker, time.Second, c.stopCh)
+
+	<-c.stopCh
+	return c.stopErr
+}
+
+func (c *UninstallController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *UninstallController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.uninstall()
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *UninstallController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	c.queue.AddRateLimited(key)
+}
+
+func (c *UninstallController) uninstall() (err error) {
+	defer func() {
+		if err != nil {
+			if c.lastErr == nil || c.lastErr.Error() != err.Error() {
```

Comment:
When do you use `lastErr`? It looks like you ignored the error and avoid passing it to `handleErr()`, why?

Reply:
`lastErr` effectively holds the last status of the controller (or, in rare events, a real error). We do a quick diff to avoid spamming the logs with no-op worker loops.

Ignoring the error must happen because every worker execution returns an error until the controller completes. This was a design change introduced in this code review to fix confusing method signatures.